C error on [1]&int{}[0]

**V version:** `V 0.3.0 3f3fb5b.819b6f4`
**OS:**  `linux, Linux version 5.18.8 (nixbld@localhost) (gcc (GCC) 11.3.0, GNU ld (GNU Binutils) 2.38) #1-NixOS SMP PREEMPT_DYNAMIC Wed Jun 29 07:04:44 UTC 2022`

While experimenting with #14958, I noticed that the V code `[1]&int{}[0]` compiles to the invalid C code `{0}[0]`, leading to “C error. This should never happen.”

```console
$ cat test.v
fn main() {
    println([1]&int{}[0])
}
$ ./v -cg run test.v
tcc: error: file 'crt1.o' not found
tcc: error: file 'crti.o' not found
/tmp/v_1000/test.3450692629302334504.tmp.c:12076: error: identifier expected
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -327,15 +327,27 @@ fn (mut g Gen) index_of_fixed_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 	elem_sym := g.table.sym(elem_type)
 	is_fn_index_call := g.is_fn_index_call && elem_sym.info is ast.FnType
 
-	if is_fn_index_call {
-		g.write('(*')
-	}
-	if node.left_type.is_ptr() {
-		g.write('(*')
+	if node.left is ast.ArrayInit {
+		tmp := g.new_tmp_var()
+		line := g.go_before_stmt(0).trim_space()
+		styp := g.typ(node.left_type)
+		g.empty_line = true
+		g.write('$styp $tmp = ')
 		g.expr(node.left)
-		g.write(')')
+		g.writeln(';')
+		g.write(line)
+		g.write(tmp)
 	} else {
-		g.expr(node.left)
+		if is_fn_index_call {
+			g.write('(*')
+		}
+		if node.left_type.is_ptr() {
+			g.write('(*')
+			g.expr(node.left)
+			g.write(')')
+		} else {
+			g.expr(node.left)
+		}
 	}
 	g.write('[')
 	direct := unsafe { g.fn_decl != 0 } && g.fn_decl.is_direct_arr

--- a/vlib/v/tests/fixed_array_literal_index_test.v
+++ b/vlib/v/tests/fixed_array_literal_index_test.v
@@ -1,0 +1,7 @@
+fn test_fixed_array_literal_index() {
+   println([1]int{}[0])
+	assert [1]int{}[0] == 0
+
+	println([1, 2]![1])
+	assert [1, 2]![1] == 2
+}

--- a/vlib/v/tests/fixed_array_literal_index_test.v
+++ b/vlib/v/tests/fixed_array_literal_index_test.v
@@ -1,5 +1,5 @@
 fn test_fixed_array_literal_index() {
-   println([1]int{}[0])
+	println([1]int{}[0])
 	assert [1]int{}[0] == 0
 
 	println([1, 2]![1])

